src/mlpack/methods/ann/dag_network_impl.hpp

```diff
@@ -0,0 +1,865 @@
+        "layers!");
+  }
+
+  if (!graphIsSet)
+    CheckGraph();
+
+  if (!validOutputDimensions)
+    UpdateDimensions(functionName, inputDimensionality);
+
+  if (parameters.is_empty())
+  {
+    InitializeWeights();
+  }
+  else if (parameters.n_elem != WeightSize())
+  {
+    parameters.clear();
+    InitializeWeights();
+  }
+
+  if (!layerMemoryIsSet)
+    SetLayerMemory();
+
+  if (setMode)
+    SetNetworkMode(training);
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::InitializeForwardPassMemory(const size_t batchSize)
+{
+  CheckNetwork("DAGNetwork::InitializeForwardPassMemory()", 0);
+
+  size_t totalOutputSize = 0;
+  for (size_t i = 0; i < sortedNetwork.size() - 1; i++)
+  {
+    totalOutputSize += sortedNetwork[i]->OutputSize();
+  }
+
+  size_t totalConcatSize = 0;
+  for (size_t i = 1; i < sortedNetwork.size(); i++)
+  {
+    if (adjacencyList[sortedNetwork[i]].size() > 1)
+    {
+      size_t concatSize = sortedNetwork[i]->InputDimensions()[0];
+      for (size_t j = 1; j < sortedNetwork[i]->InputDimensions().size(); j++)
+      {
+        concatSize *= sortedNetwork[i]->InputDimensions()[j];
+      }
+      totalConcatSize += concatSize;
+    }
+  }
+
+  size_t forwardMemSize = totalOutputSize + totalConcatSize;
+  if (batchSize * forwardMemSize > layerOutputMatrix.n_elem)
+  {
+    layerOutputMatrix = MatType(1, batchSize * forwardMemSize);
+  }
+
+  size_t offset = 0;
+  sortedIndices.clear();
+  //setup layerOutputs
+  for (size_t i = 0; i < sortedNetwork.size() - 1; i++)
+  {
+    const size_t layerOutputSize = sortedNetwork[i]->OutputSize();
+    MakeAlias(layerOutputs[i], layerOutputMatrix, layerOutputSize, batchSize, offset);
+    offset += batchSize * layerOutputSize;
+    sortedIndices.insert({sortedNetwork[i], i});
+  }
+
+  //setup layerInputs
+  for (size_t i = 1; i < sortedNetwork.size(); i++)
+  {
+    Layer<MatType>* currentLayer = sortedNetwork[i];
+    std::vector<Layer<MatType>*> parents = adjacencyList[currentLayer];
+    size_t numParents = parents.size();
+    if (numParents <= 0)
+    {
+      // TODO: should be caught in CheckGraph
+      throw std::logic_error("DAGNetwork::InitializeForwardPassMemory(): Non-input nodes must have atleast one parent node.");
+    }
+
+    if (numParents == 1)
+    {
+      size_t inputIndex = sortedIndices[parents.front()];
+      MakeAlias(layerInputs[i-1], layerOutputs[inputIndex], layerOutputs[inputIndex].n_rows, layerOutputs[inputIndex].n_cols, 0);
+    }
+    else
+    {
+      size_t concatSize = currentLayer->InputDimensions()[0];
+      for (size_t j = 1; j < currentLayer->InputDimensions().size(); j++)
+        concatSize *= currentLayer->InputDimensions()[j];
+      MakeAlias(layerInputs[i-1], layerOutputMatrix, concatSize, batchSize, offset);
+      offset += concatSize;
```

Comment:
What happens if the concat axis is not the last axis?  This code will work for that, but I don't think that it will in any other case.  (Even if the answer will become clear to me as I keep reading, it wouldn't be a bad idea to add a comment here indicating that other concat axes are handled later, and we still just use adjacent memory for the direct layer outputs before we interleave them.)